core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java

```diff
@@ -0,0 +1,111 @@
+    private final Runnable setWritersFullyScaled;
+
+    // start with single writer and increase the writer count as buffer memory utilization increases
+    private int writerCount = 1;
+    private int nextWriterIndex = -1;
+    private long lastScaleUpPhysicalWrittenBytes;
+    private boolean isSetWritersFullyScaledExecuted;
+
+    public ScaleWriterExchanger(
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Supplier<Long> physicalWrittenBytesSupplier,
+            DataSize writerMinSize,
+            Runnable setWritersFullyScaled)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.onPageReleased = PageReleasedListener.forLocalExchangeMemoryManager(memoryManager);
+        // We are aiming to keep the buffer size below 50% of its limit
+        this.bufferSizeMinScalingThreshold = maxBufferedBytes / 2;
+        this.physicalWrittenBytesSupplier = requireNonNull(physicalWrittenBytesSupplier, "physicalWrittenBytesSupplier is null");
+        this.writerMinSize = requireNonNull(writerMinSize, "writerMinSize is null").toBytes();
+        this.setWritersFullyScaled = requireNonNull(setWritersFullyScaled, "setWritersFullyScaled is null");
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Consumer<PageReference> buffer = buffers.get(getNextWriterIndex());
+        PageReference pageReference = new PageReference(page, 1, onPageReleased);
+        memoryManager.updateMemoryUsage(pageReference.getRetainedSizeInBytes());
+        buffer.accept(pageReference);
+    }
+
+    private int getNextWriterIndex()
+    {
+        // Scale up writers when current buffer memory utilization is more than 50% of
+        // maximum and physical written bytes by the last scaled up writer is greater than
+        // writerMinSize.
+        if (writerCount < buffers.size() && memoryManager.getBufferedBytes() >= bufferSizeMinScalingThreshold) {
+            long physicalWrittenBytes = physicalWrittenBytesSupplier.get();
+            if ((physicalWrittenBytes - lastScaleUpPhysicalWrittenBytes) >= writerCount * writerMinSize) {
+                lastScaleUpPhysicalWrittenBytes = physicalWrittenBytes;
+                writerCount++;
+                log.debug("Increased task writer count: %d", writerCount);
+            }
+        }
+
+        // Once writers are fully scaled, it will update its status which further tells
+        // the scheduler to add another task if scale_writers is enabled.
+        if (!isSetWritersFullyScaledExecuted && writerCount == buffers.size()) {
+            long physicalWrittenBytes = physicalWrittenBytesSupplier.get();
+            if ((physicalWrittenBytes - lastScaleUpPhysicalWrittenBytes) >= writerCount * writerMinSize) {
+                setWritersFullyScaled.run();
```

Comment:
If my understanding is correct a new instance of `ScaleWriterExchanger` is created for every upstream driver. For simple queries like `INSERT INTO ... SELECT ... FROM ...` a new driver is created for every split. In other words `ScaleWriterExchanger` is not long lived. 

There are two issues that I can see associated with that.

1. For simple `INSERT from SELECT` queries writers are unlikely to scale beyond two, as initially `physicalWrittenBytes - 0` will pass, but then the same condition is unlikely to pass within a single split.
2. As a consequence of first it is unlikely the writers will ever be fully "scalled" what could prevent distributed scaling.

Reply:
> For simple INSERT from SELECT queries writers are unlikely to scale beyond two, as initially physicalWrittenBytes - 0 will pass, but then the same condition is unlikely to pass within a single split.

Here we are checking `physicalWrittenBytes` at a task level, not a driver level. For instance, let's say, there are 6 drivers and 6 instances of `ScaleWriterExchanger`, so if any of the instances reaches the maximum, we will scale globally.

That's my understanding. Please correct me if I'm wrong.

Reply:
I'm trying to think about the following scenario:

0. Some number of splits have finished
1. A new split is scheduled.
2. `physicalWrittenBytes - lastScaleUpPhysicalWrittenBytes` passes the first time, as `physicalWrittenBytes - 0` will likely be greater than threshold
3. A new writer is added
4. The split finishes before  `physicalWrittenBytes - lastScaleUpPhysicalWrittenBytes` passes again

Reply:
What if we use the same instance of `ScaleWriterExchanger` for a task? I mean, I'll benchmark this but just an idea.

Reply:
We can have a singleton object that used by all `ScaleWriterExchanger` created that stores the scalling related state and the current desired number of writers.

Reply:
This is done